### PR TITLE
openjdk21-graalvm: update minimum OS version

### PR DESCRIPTION
**Pull request by breun**
#### Description

Correct the minimum OS version. Also see https://trac.macports.org/ticket/71045.

###### Type(s)

- [x] bugfix

###### Tested on

macOS 15.0.1 24A348 arm64
Xcode 16.0 16A242d

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: java/openjdk21-graalvm/Portfile [hunk 1]
```diff
@@ -2,10 +2,16 @@
...
+set feature 21
+name             openjdk${feature}-graalvm
 categories       java devel
 maintainers      {breun.nl:nils @breun} openmaintainer
-platforms        {darwin any}
+
+# JVMMinimumSystemVersion in Contents/Info.plist is set to macOS 11.00.00:
+# /usr/libexec/PlistBuddy -c "Print :JavaVM:JVMMinimumSystemVersion" Contents/Info.plist
+# Mapping to Darwin version: https://trac.macports.org/wiki/PortfileRecipes#compare-osx-darwin-version
+platforms        {darwin any >= 20 }
```
Comment (sambthompson) [inline java/openjdk21-graalvm/Portfile:13]: Are you sure about this? My install on 10.14.6 reports min ver as 10.12.0?

Comment (breun) [reply java/openjdk21-graalvm/Portfile:13]: This is what I see:

```
❯ /usr/libexec/PlistBuddy -c "Print :JavaVM:JVMMinimumSystemVersion" /Library/Java/JavaVirtualMachines/jdk-21-oracle-graalvm-community.jdk/Contents/Info.plist 
11.00.00
```

Do you get 10.12.0?

Comment (breun) [reply java/openjdk21-graalvm/Portfile:13]: Ah, I guess I should check the x86_64 tarballs...

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: Yes, I'm on x86_64; I imagine the minimum for arm64 would be different.

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: Probably affects openjdk17-graalvm (which is what led me to check, since I had later version installed no issue). Also 10.12.0 reported for jdk-21-oracle-graalvm.jdk and jdk-21-oracle-openjdk.jdk.

Comment (breun) [reply java/openjdk21-graalvm/Portfile:13]: Yes, I'll have to re-check all the changes I made now...

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: Hopefully lower min vers for x64_86 won't cause issues for arm users; I don't know if Portfiles support different min vers on different architectures?

Comment (breun) [reply java/openjdk21-graalvm/Portfile:13]: There won't be any arm64 users on macOS 10.x, because macOS 11 was the first version with an arm64 variant.

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: What I meant was, if the JDK specifies a higher min version than 11 for arm64, but some 10.x still OK for x86_64. Very unlikely, I guess. Anyway, thanks again for all this clean-up!

Comment (breun) [reply java/openjdk21-graalvm/Portfile:13]: We'll cross that bridge when we get there. (But I indeed expect we won't.)

Would you be able to check the minimum versions on x86_64 for the `openjdk8`, `openjdk11`, `openjdk17`, `openjdk21` and `openjdk22` ports that are built from source for me?

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: Let me have a look. I seem to recall getting a build failure on one of these which is when I started with the packaged alternatives and here we are. But I'll need to re-run.

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: Wouldn't ports built from source either enforce the limits (and fail to build) or just set min ver to whatever build platform is when generating the plist?

Comment (breun) [reply java/openjdk21-graalvm/Portfile:13]: I don't know if the contents of `Contents/Info.plist` depend on the OS version of the build machine. I would expect this to depend on the architecture of the build machine only, but I'm not sure.

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: I would hope so. If that's the case, it may be possible to determine that from the source at the completion of the configure phase. Just having a look at `openjdk8` now.

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: Yes, the build sets the minimum version and it's specified in the openjdk sources (subject to target arch as you suggested), and not from the machine OS, or Portfile.

tl;dr:
{{{
openjdk8  10.6.0
openjdk11 10.12.0
openjdk17 10.12.0
openjdk21 10.12.0
openjdk22 11.00
}}}

So, for `openjdk8`, after `port build configure`, identified bundle Info-plist templates in the source. Running `/usr/libexec/PlistBuddy -c "Print :JavaVM:JVMMinimumSystemVersion" /opt/local/var/macports/build/_opt_local_var_macports_sources_github.com_macports_macports-ports_java_openjdk8/openjdk8/work/jdk8u-jdk8u372-ga/jdk/src/macosx/bundle/JDK-Info.plist` returns `10.6.0`.

Looking for these templates in the git repo, identified that later versions don't hard code. The variable `MACOSX_VERSION_MIN` set in `make/autoconf/flags.m4`, has been set since March 2018, initially to `10.7`, which should cover `openjdk11`, but the value in the template plist remained hard-coded in builds until openjdk16, when [this change](https://github.com/openjdk/jdk/commit/7df86108cb42079df8ade07b4d07587c1f0cbe83#diff-5930e43f9578abc410471f7cc5fa7e4ec288451979d2d764f6644b7f4d2fea9a) landed.

This suggests an early build of `openjdk11` would set `10.6` as minimum, but the correct value should have been `10.7` but this was not enforced in the bundle. I assume that means a build on `10.6` would fail, but using a build from `>10.6` on `10.6` wouldn't be detected at run-time (but presumably there is some feature missing on 10.6 that would cause crash, etc. However, current builds of `openjdk11` have the variable in the template, so change to 10.12.0 in 2021 is enforced.

For `openjdk17` thru openjdk21 (for `x86_64`, once `arm64` added), this has been set to `10.12` since 2018.
For `openjdk22`, all arches set to `11.00`. Identifying the minimum version should not require a full build, but running `grep -C2 "MACOSX_VERSION_MIN=" /opt/local/var/macports/build/*openjdk*/openjdk*/work/jdk-*/make/autoconf/flags.m4` after `port configure` should do it.

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: Per the tags on [this commit](https://github.com/openjdk/jdk/commit/72c4dcbfeefcd664f5e3175b24e395c1f36a05fd), bump to 11.00 wasn't applied to openjdk21 for all arch's. I think that means #26097 has min version set too high, per above analysis. I have checked this with steps listed:

```
/opt/local/var/macports/build/_opt_local_var_macports_sources_github.com_macports_macports-ports_java_openjdk21/openjdk21/work/jdk-21.0.4+7/make/autoconf/flags.m4
-    # require code changes to build.
-    if test "x$OPENJDK_TARGET_CPU_ARCH" = xaarch64; then
:      MACOSX_VERSION_MIN=11.00.00
-    else
:      MACOSX_VERSION_MIN=10.12.0
-    fi
-    MACOSX_VERSION_MIN_NODOTS=${MACOSX_VERSION_MIN//\./}
```

Comment (breun) [reply java/openjdk21-graalvm/Portfile:13]: It may indeed be possible to build OpenJDK 21 on some macOS 10.x versions, but according to https://ports.macports.org/port/openjdk21/details/ builds were failing on all macOS 10.x versions. I haven't looked into the cause of those failures yet.

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: Flagging this discussion in remedial PRs: #26092, #26093, #26094, #26095, #26096, #26098.

Comment (breun) [reply java/openjdk21-graalvm/Portfile:13]: The log for Catalina (https://build.macports.org/builders/ports-10.15_x86_64-builder/builds/196610/steps/install-port/logs/stdio) shows:

```
src/java.desktop/macosx/native/libawt_lwawt/awt/CGraphicsDevice.m:69:29: error: use of undeclared identifier 'NSBundleExecutableArchitectureARM64'; did you mean 'NSBundleExecutableArchitecturePPC64'?
        if (architecture == NSBundleExecutableArchitectureARM64) {
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                            NSBundleExecutableArchitecturePPC64
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSBundle.h:123:5: note: 'NSBundleExecutableArchitecturePPC64' declared here
    NSBundleExecutableArchitecturePPC64     = 0x01000012
    ^
1 error generated.
```

I'm not surprised that a pre-ARM64 macOS like Catalina doesn't know about something called `NSBundleExecutableArchitectureARM64`, but I don't really know how to fix this.

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: > It may indeed be possible to build OpenJDK 21 on some macOS 10.x versions, but according to https://ports.macports.org/port/openjdk21/details/ builds were failing on all macOS 10.x versions. I haven't looked into the cause of those failures yet.

Ah, I see. This is probably the build that failed for me originally. I will try a build and see if the errors are already in an existing ticket, or open a new one as appropriate.

Comment (sambthompson) [reply java/openjdk21-graalvm/Portfile:13]: Opened https://trac.macports.org/ticket/71049 for this build failure. Probably should move further discussion to ticket or #26097.
